modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java

```diff
@@ -170,8 +183,17 @@ private IgnoreNoResponseMetricsCollector(Operation operation, OperationPurpose p
         public final void collectMetrics(Request<?> request, Response<?> response) {
             if (response != null) {
                 assert assertConsistencyBetweenHttpRequestAndOperation(request, operation);
-                counter.add(getRequestCount(request));
-                requestCounter.incrementBy(getRequestCount(request), attributes);
+                final TimingInfo timingInfo = request.getAWSRequestMetrics().getTimingInfo();
+                final long requestCount = getCountForMetric(timingInfo, AWSRequestMetrics.Field.RequestCount);
+                final long exceptionCount = getCountForMetric(timingInfo, AWSRequestMetrics.Field.Exception);
+                final long throttleCount = getCountForMetric(timingInfo, AWSRequestMetrics.Field.ThrottleException);
+
+                counter.add(requestCount);
+                requestCounter.incrementBy(requestCount, attributes);
+                exceptionCounter.incrementBy(exceptionCount, attributes);
+                throttleCounter.incrementBy(throttleCount, attributes);
+                exceptionHistogram.record(exceptionCount, attributes);
+                throttleHistogram.record(throttleCount, attributes);
```

Comment:
I think we should introduce 2 additional counters here:
* operation count
* successful or unsuccessful operation count

Operation counter is incremented by 1 here. Each operation might be backed by multiple requests (if backing http requests are redirected or retried due to the 429 or 500).

Unsuccessful operation counter should be incremented if operation exhausted all retries on 429/500/etc.
Possibly we have something in `request.getAWSRequestMetrics()` for that already.

WDYT?

Reply:
Yeah those seem to be useful. For unsuccessful operations, I compute them as "number of error code == number of requests".

Reply:
According to `com.amazonaws.http.AmazonHttpClient.RequestExecutor#captureExceptionMetrics` it looks like they increment both Exception and ThrottleException for throttle events so `number of exceptions == number of requests` should work, unless there could be a situation where a client follows a redirect (assuming it increment number of requests) but all actual requests after redirect are failing. Possibly they have some better indication in recorded metrics if request have ultimately failed.

Reply:
Not every throttle exception is an exception. I am pretty sure about it. The `executeOneRequest` call has to [throw an exception for it to be counted as an exception](https://github.com/aws/aws-sdk-java/blob/e44143f09d8cc40a395270c8551930e8ee39e1ce/aws-java-sdk-core/src/main/java/com/amazonaws/http/AmazonHttpClient.java#L1163-L1176). The `executeOneRequest` [internally performs retry](https://github.com/aws/aws-sdk-java/blob/e44143f09d8cc40a395270c8551930e8ee39e1ce/aws-java-sdk-core/src/main/java/com/amazonaws/http/AmazonHttpClient.java#L1387C14-L1387C14) as well and these retries are not exceptions. But they can be throttle exceptions.

For example, you can have an unsuccessful request that has 5 requests, 4 retries, 0 exception. In contrast, the `AWSErrorCode` is captured every time a request errors. So if these two have the same value. The request ultimately failed.

Reply:
Based on the code of `AmazonHttpClient`, it seems [redirect is handled as retry](https://github.com/aws/aws-sdk-java/blob/e44143f09d8cc40a395270c8551930e8ee39e1ce/aws-java-sdk-core/src/main/java/com/amazonaws/http/AmazonHttpClient.java#L1403-L1415). So I think we should be covering this already.